db/column_family.cc

```diff
@@ -929,8 +934,9 @@ bool ColumnFamilyData::NeedsCompaction() const {
 
 Compaction* ColumnFamilyData::PickCompaction(
     const MutableCFOptions& mutable_options, LogBuffer* log_buffer) {
+  SequenceNumber oldest_mem_seqno = super_version_->GetEarliestMemTableSequenceNumber();
```

Comment:
This approach is complicated. Instead of remembering earliest sequence number in memtable the moment when the previous flush happens in Version, it passes the recent one when PickCompaction happens. Do you have a proof that it would be the same value? My worry is that, if there is a concurrent flush going on as this PickCompaction(), is it possible that the memtable sequence number has been updated but PickCompaction() still can't see the newly flushed file?
Generally, I see it safer if the earliest memtable sequence number is remembered in Version when a new Version is generated.

Reply:
`PickCompaction` is called with `DB::mutex_` held, right? If there is a concurrent flush going on the memtables are either still present in `super_version_->imm`, or the flush has completed in which case it will have installed in `super_version_->current`. It looks like `ColumnFamily::current_` and `ColumnFamily::super_version_` are both updated with `DB::mutex_` held and should be identical. I'm not seeing what would be different about the approach taken in this PR and memoizing the earliest unflushed sequence number when the version is installed. I'm just reading code, though. Very possible that I'm missing something. 

Is there always an earliest memtable sequence number when a new version is installed. If I write a single key to the DB and call `Flush`, a new version will be created, and a new memtable will be created, but there won't be an earliest memtable sequence number.

Reply:
I do not have a proof, but my reading of the code suggests the following sequence of events in a single thread:

- `LogAndApply()` is called on one or more memtables that have finished flushing.
  - `ColumnFamilyData::current_` is updated which will include L0 files corresponding to the flushed memtables.
- The flushed memtables are removed from the `MemTableListVersion`.
- `InstallSuperVersion()` writes a new `ColumnFamilyData::super_version_` with the updated memtable list that excludes the flushed memtables.

Now, when this sequence of events happens concurrently, I believe the problematic sequence you mentioned should not happen.  In particular because a flushed L0 file is installed to `ColumnFamilyData::current_` *before* its corresponding memtable is dropped from `ColumnFamilyData::super_version_`. For a given memtable, these two updates happen in the same thread, so even if lock is released between the two updates, it should not be possible for another thread to drop our memtable in `super_version_`.

Reply:
These operations all happens during held mutex. And even if we get earliest seqno from an older  super_version, it does not matter. Because smaller seqno would pick less file. We only need to wait for next time to pick these files.

Reply:
Can we directly use ColumnFamilyData::mem_, ColumnFamilyData::imm_ to calculate the value instead of getting from SuperVersion? If everything is under DB mutex, then current_, mem_ and imm_ should all be from the same snapshot. Does it work?

Reply:
Yes, it works. But  there has been an available interface of SuperVersion. I just can not understand why we do not use it. I thought `ColumnFamilyData` would call `InstallSuperVersion` every time when switch memtable or flush job was finished. Is it right?

Reply:
I am not 100% sure about this. I would need to check source code to double check. When we rely on such an assumption that is applicable to change in the future, we are making the code prone to bugs in the future. Another reason is, like what you have said, SuperVersion is created after updating current_, imm_, or mem_ but it doesn't have to be atomic to it. It's perfectly OK to update current_, imm_ or mem_, release the DB mutex, do something like and then install super version. So SuperVersion is not guaranteed to be in sync with imm_ and mem_. I saw the argument that a stale SuperVersion should still be OK in this case, the argument seems to be plausible to me, but if we don't have to rely on such proof to make sure the code is correct, it is even better.